src/js/utils/validator.js

```diff
@@ -1,44 +1,106 @@
+        this.errors[fieldName] = `The ${this.getFriendlyName(fieldName)} is required`;
+      }
+    }
+  }
 
-/**
- * @desc validate data with email and required field check.
- * @param {object} data Data as represented in the component state.
- * @param {array} optionalFields The input name as represented in the state of
- *        the component as key. For example, if the message field in a form is
- *        optional and the data is {message: 'Hello'}, the optional fields
- *        should be ['message'].
- * @returns {object} The result containing errors, formatted errors and a
- *        boolean to show validity.
- */
-function validate(data, optionalFields) {
-  let errors = {};
-
-  if(!isValidEmail(data.email)) {
-    errors.email = `The email is invalid`;
-  }
-
-  for(let key in data) {
-    if(!optionalFields || optionalFields.indexOf(key) === -1) {
-      if(typeof data[key] == 'string' && isEmptyString(data[key])) {
-        errors[key] = `The ${key} is required`;
+  testMaxLength(validatorOption, fieldName) {
+    if (this.data[fieldName].toString().length < validatorOption) {
+      if (!this.errors[fieldName]) {
+        this.errors[fieldName] = `The ${this.getFriendlyName(fieldName)} must be up to ${validatorOption} characters long`;
       }
     }
   }
 
-  return {
-    errors,
-    concatenatedErrors: Object.values(errors).join('. '),
-    isValid: Object.keys(errors).length === 0
-  };
+  testEmail(fieldName) {
+    if (!this.validateEmail(this.data[fieldName])) {
+      this.errors[fieldName] = this.defaultErrorMessages[fieldName];
+    }
+  }
+
+  testPhoneNumber(fieldName) {
+    if (!this.validatePhoneNumber(this.data[fieldName])) {
+      this.errors[fieldName] = this.defaultErrorMessages[fieldName];
+    }
+  }
+
+  validatorResults() {
+    return {
+      errors: this.errors,
+      concatenatedErrors: Object.values(this.errors).join('. '),
+      isValid: Object.keys(this.errors).length === 0
+    };
+  }
+
+  validateByOptions() {
+    Object.keys(this.options).forEach(fieldName => {
+      Object.keys(this.options[fieldName]).forEach(optionKey => {
+        let validatorOption = this.options[fieldName][optionKey];
+
+        switch(optionKey) {
+        case "message": {
+          this.defaultErrorMessages[fieldName] = validatorOption;
+          break;
+        }
+        case "required": {
+          this.testRequired(validatorOption, fieldName);
+          break;
+        }
+        case "maxLength": {
+          this.testMaxLength(validatorOption, fieldName);
+          break;
+        }
+        }
+      })
+    })
+  }
+
+  validateAllInputs() {
+    if(!this.data || Object.keys(this.data).length === 0) {
+      return this.validatorResults();
+    }
+
+    this.testEmail('email');
+    this.data.phone_number && this.testPhoneNumber('phone_number');
+    if (this.options) {
+      this.validateByOptions(this.data, this.options);
+    }
+
+    return this.validatorResults();
```

Comment:
Great job with the validators! The class still does a lot, but for now, behavior is better encapsulated. We'll leave further improvements as technical debt. For now, this can fly. ❤️